File: backend/app/schemas/tag.py
"""Tag registry schemas — 测点清单 (IDS §测点管理)."""

from __future__ import annotations

from pydantic import Field

from app.schemas.base import CamelModel


class TagLoopInfo(CamelModel):
    """测点关联的回路信息（通过 loop_tag_mapping 间接关联）。"""

    loopId: str
    loopTagName: str
    loopDescription: str | None = None


class TagListItem(CamelModel):
    """测点列表项。"""

    id: str
    tagName: str
    tagDescription: str | None = None
    tagType: str
    currentValue: float | None = None
    quality: str | None = None
    lastSyncAt: str | None = None
    isLinked: bool | None = None
    rangeMin: float | None = None
    rangeMax: float | None = None
    unit: str | None = None
    measureType: str | None = None
    tdengineTagId: str | None = None
    loop: TagLoopInfo | None = None


class TagListData(CamelModel):
    """测点列表响应 data 块。"""

    items: list[TagListItem]
    total: int
    page: int
    pageSize: int


class TagDetail(CamelModel):
    """测点详情。"""

    id: str
    tagName: str
    tagDescription: str | None = None
    tagType: str
    currentValue: float | None = None
    quality: str | None = None
    lastSyncAt: str | None = None
    isLinked: bool | None = None
    rangeMin: float | None = None
    rangeMax: float | None = None
    unit: str | None = None
    measureType: str | None = None
    tdengineTagId: str | None = None
    loop: TagLoopInfo | None = None


class TagUpdate(CamelModel):
    """PUT /api/v1/tags/{id} 请求体。"""

    tagDescription: str | None = Field(None, max_length=255)
    rangeMin: float | None = None
    rangeMax: float | None = None
    unit: str | None = Field(None, max_length=20)
    measureType: str | None = None
    tdengineTagId: str | None = Field(None, max_length=100)


class TagDeleteResult(CamelModel):
    """测点删除响应。"""

    id: str
    deleted: bool = True
    deletedAt: str


class TagBatchDeleteRequest(CamelModel):
    """批量删除测点请求体。"""

    tagIds: list[str] = Field(..., min_length=1, max_length=500)


class TagBatchDeleteFailure(CamelModel):
    """批量删除中单个失败项。"""

    tagId: str
    tagName: str | None = None
    reason: str


class TagBatchDeleteResult(CamelModel):
    """批量删除测点响应。"""

    deleted: int
    failed: int
    failures: list[TagBatchDeleteFailure] = []


class TagImportError(CamelModel):
    """测点导入单行错误。"""

    row: int
    tagName: str | None = None
    message: str


class TagImportResult(CamelModel):
    """POST /api/v1/tags/import 响应."""

    total: int
    inserted: int
    updated: int
    failed: int
    errors: list[TagImportError] = []


# ---------------------------------------------------------------------------
# v4.0 波形数据 schema（IDS §2.4.5 — 扩展 valid_mask + tagGroup 筛选）
# 设计依据：IDS §2.4.5, 算法说明 §3.4（KEEP_ALL_WITH_VALIDITY）/§3.7.1（数据血缘）
# ---------------------------------------------------------------------------


class WaveformTimeRange(CamelModel):
    """波形时间范围."""

    startTime: str
    endTime: str


class WaveformPoint(CamelModel):
    """波形数据点（含 valid_mask 标记）.

    设计依据：IDS §2.4.5, 算法说明 §3.4（KEEP_ALL_WITH_VALIDITY 策略）

    Attributes:
        timestamp: ISO 8601 时间戳
        pv: 过程值
        sp: 设定值
        op: 操作输出
        mode: 控制模式（0=手动, 1=自动）
        pvQuality: PV 质量码（1=Good, 0=Bad）
        valid: valid_mask 标记（True=有效, False=无效/异常）
        outlierReason: 异常原因码（如 FROZEN/JUMP/SPIKE/OUT_OF_RANGE 等，多个以逗号分隔）
    """

    timestamp: str
    pv: float | None = None
    sp: float | None = None
    op: float | None = None
    mode: int | None = None
    pvQuality: int | None = None
    valid: bool = True
    outlierReason: str | None = None


class WaveformResponse(CamelModel):
    """波形响应（含 valid_mask + 数据血缘）.

    设计依据：IDS §2.4.5, 算法说明 §3.7.1

    Attributes:
        loopId: 回路 ID
        tagName: 回路位号
        timeRange: 时间范围
        points: 波形数据点列表（含 valid_mask）
        samplingFreq: 采样频率（如 ``1s`` / ``5s``）
        qualityPolicy: 质量策略（``KEEP_ALL_WITH_VALIDITY`` / ``KEEP_ALL``）
        validRate: 有效数据率（0~1）
        downsampled: 是否经过 LTTB 降采样
        pointCount: 返回的数据点数
    """

    loopId: str
    tagName: str | None = None
    timeRange: WaveformTimeRange
    points: list[WaveformPoint] = Field(default_factory=list)
    samplingFreq: str = "1s"
    qualityPolicy: str = "KEEP_ALL_WITH_VALIDITY"
    validRate: float = 1.0
    downsampled: bool = False
    pointCount: int = 0


class BatchWaveformRequest(CamelModel):
    """POST /api/v1/timeseries/batch/waveform 请求体.

    批量查询多个回路的波形数据，使用 ``asyncio.gather`` 并行获取。

    Attributes:
        loopIds: 回路 ID 列表（1~50 个）
        startTime: 开始时间（ISO 8601）
        endTime: 结束时间（ISO 8601）
        tagGroup: 按标签组筛选（BASE/OP_HF/PVOP_HF/MODE_HF/QUALITY_HF），默认 BASE
        includeValidMask: 是否返回 valid_mask（默认 True）
        maxPoints: 每个回路最大数据点数（100~50000，默认 5000）
    """

    loopIds: list[str] = Field(..., min_length=1, max_length=50)
    startTime: str
    endTime: str
    tagGroup: str | None = Field(
        None, description="按标签组筛选: BASE/OP_HF/PVOP_HF/MODE_HF/QUALITY_HF"
    )
    includeValidMask: bool = True
    maxPoints: int = Field(5000, ge=100, le=50000)


class BatchWaveformFailure(CamelModel):
    """批量查询中失败的回路信息."""

    loopId: str
    error: str


class BatchWaveformResponse(CamelModel):
    """批量波形查询响应.

    Attributes:
        items: 成功获取的波形数据列表
        failed: 失败的回路列表（含错误信息）
        total: 成功回路数
    """

    items: list[WaveformResponse] = Field(default_factory=list)
    failed: list[BatchWaveformFailure] = Field(default_factory=list)
    total: int = 0


__all__ = [
    "BatchWaveformFailure",
    "BatchWaveformRequest",
    "BatchWaveformResponse",
    "TagDeleteResult",
    "TagDetail",
    "TagImportError",
    "TagImportResult",
    "TagListData",
    "TagListItem",
    "TagLoopInfo",
    "TagUpdate",
    "WaveformPoint",
    "WaveformResponse",
    "WaveformTimeRange",
]

File: frontend/apps/web-antd/src/api/tag.ts
/**
 * CLPM 测点清单 API
 *
 * 覆盖测点列表查询、详情、更新、删除及批量导入导出能力。
 */
import type { PageQuery, PaginatedResponse } from '#/api/types';

import { requestClient } from '#/api/request';

export namespace TagApi {
  /** 测点类型（温度/压力/液位/流量/分析/速度/其他） */
  export type MeasureType =
    | 'ANALYSIS'
    | 'FLOW'
    | 'LEVEL'
    | 'OTHER'
    | 'PRESSURE'
    | 'SPEED'
    | 'TEMPERATURE';

  /** 参数类型（PV/SP/OP/MODE/PID_P/PID_I/PID_D/OTHER） */
  export type TagType =
    | 'MODE'
    | 'OP'
    | 'OTHER'
    | 'PID_D'
    | 'PID_I'
    | 'PID_P'
    | 'PV'
    | 'SP';

  /** 质量戳（GOOD/BAD/UNCERTAIN） */
  export type Quality = 'BAD' | 'GOOD' | 'UNCERTAIN';

  /** 测点列表项 */
  export interface TagItem {
    id: string;
    tagName: string;
    tagDescription?: string;
    tagType: TagType;
    currentValue?: null | number;
    quality?: null | Quality;
    lastSyncAt?: null | string;
    isLinked: boolean;
    rangeMin?: null | number;
    rangeMax?: null | number;
    unit?: string;
    measureType?: MeasureType | null;
    tdengineTagId?: string;
    loopId?: string;
    loopTagName?: string;
    loopDescription?: string;
    unitName?: string;
    /**
     * R06（数据链路整改）：实时新值无效（"-1.#QNAN0"/"nan"/"Infinity"/空值）时
     * 后端返回 currentValue=null 且 stale=true（不以 DB 旧值伪装最新有效读数）；
     * 前端 WS 失联期间也本地置位标旧
     */
    stale?: boolean;
    /**
     * 数据健康度（方案 C 轻量版）：实时质量码 + 同步新鲜度 + 所属回路 PV 完整度
     * 来自每日巡检快照，不在列表页逐 tag 实时查 TDengine
     */
    dataHealth?: TagDataHealth;
  }

  /** 测点数据健康度 */
  export interface TagDataHealth {
    /** 实时质量码（GOOD/BAD/UNCERTAIN） */
    quality?: null | Quality;
    /** 同步新鲜度（最近落库时间 ISO 串） */
    lastSyncAt?: null | string;
    /** 所属回路 PV 完整度（0~1，来自每日巡检快照） */
    loopPvCompleteness?: null | number;
    /** 所属回路完整性状态：OK/WARNING/CRITICAL/DATA_UNAVAILABLE */
    loopIntegrityStatus?: null | string;
    /** 最近巡检日期 */
    lastIntegrityCheck?: null | string;
  }

  /** 测点列表查询参数 */
  export interface TagQueryParams extends PageQuery {
    keyword?: string;
    /** 测点类型字典 code（可配置：系统管理 → 字典管理） */
    measureType?: string;
    /** 参数类型字典 code（可配置：系统管理 → 字典管理） */
    tagType?: string;
    plantNodeId?: string;
    isLinked?: boolean;
  }

  /** 更新测点参数 */
  export interface UpdateTagParams {
    tagDescription?: string;
    rangeMin?: null | number;
    rangeMax?: null | number;
    unit?: string;
    /** 测点类型字典 code（可配置：系统管理 → 字典管理） */
    measureType?: string;
    /** 参数类型字典 code（可配置：系统管理 → 字典管理） */
    tagType?: string;
    tdengineTagId?: string;
  }

  /** 新建测点参数 */
  export interface CreateTagParams {
    tagName: string;
    tagDescription?: string;
    rangeMin?: null | number;
    rangeMax?: null | number;
    unit?: string;
    /** 测点类型字典 code（可配置：系统管理 → 字典管理） */
    measureType?: string;
    /** 参数类型字典 code（可配置：系统管理 → 字典管理） */
    tagType?: string;
    tdengineTagId?: string;
  }

  /** Excel 导入单行错误 */
  export interface TagImportError {
    row: number;
    tagName?: null | string;
    message: string;
  }

  /** Excel 导入结果（upsert：位号存在则更新，否则新建） */
  export interface TagImportResult {
    /** 总行数 */
    total: number;
    /** 新建数 */
    inserted: number;
    /** 更新数 */
    updated: number;
    /** 失败数 */
    failed: number;
    /** 失败行明细 */
    errors: TagImportError[];
  }
}

/**
 * 获取测点列表（分页）
 */
export function getTagListApi(params: TagApi.TagQueryParams) {
  return requestClient.get<PaginatedResponse<TagApi.TagItem>>('/tags', {
    params,
  });
}

/**
 * 获取测点详情
 */
export function getTagDetailApi(tagId: string) {
  return requestClient.get<TagApi.TagItem>(`/tags/${tagId}`);
}

/**
 * 新建测点（ADMIN/IC_ENGINEER）
 *
 * 位号唯一；isLinked 恒为 false（仅由回路映射派生）
 */
export function createTagApi(data: TagApi.CreateTagParams) {
  return requestClient.post<TagApi.TagItem>('/tags', data);
}

/**
 * 更新测点
 */
export function updateTagApi(tagId: string, data: TagApi.UpdateTagParams) {
  return requestClient.put<TagApi.TagItem>(`/tags/${tagId}`, data);
}

/**
 * 删除测点
 */
export function deleteTagApi(tagId: string) {
  return requestClient.delete<null>(`/tags/${tagId}`);
}

/** 批量删除失败项 */
export interface TagBatchDeleteFailure {
  tagId: string;
  tagName?: string;
  reason: string;
}

/** 批量删除结果 */
export interface TagBatchDeleteResult {
  deleted: number;
  failed: number;
  failures: TagBatchDeleteFailure[];
}

/**
 * 批量删除测点（仅 ADMIN）
 *
 * 已关联回路的测点跳过并记入 failures。
 */
export function batchDeleteTagsApi(tagIds: string[]) {
  return requestClient.post<TagBatchDeleteResult>('/tags/batch-delete', {
    tagIds,
  });
}

/**
 * 根据回路位号自动匹配测点
 */
export function matchTagsForLoopApi(loopTagName: string) {
  return requestClient.get<
    Array<{
      measureType?: string;
      role: string;
      tagDescription?: string;
      tagId: string;
      tagName: string;
      tagType: string;
      unit?: string;
    }>
  >('/tags/match-loop', {
    params: { loopTagName },
  });
}
